mlir/lib/Dialect/SCF/IR/SCF.cpp

```diff
@@ -1509,6 +1510,203 @@ class ForallOpControlOperandsFolder : public OpRewritePattern<ForallOp> {
+///  INPUT:
+///   %res:3 = scf.forall ... shared_outs(%arg0 = %a, %arg1 = %b, %arg2 = %c)
+///            {
+///                ...
+///                <SOME USE OF %arg0>
+///                <SOME USE OF %arg1>
+///                <SOME USE OF %arg2>
+///                ...
+///                scf.forall.in_parallel {
+///                    <STORE OP WITH DESTINATION %arg1>
+///                    <STORE OP WITH DESTINATION %arg0>
+///                    <STORE OP WITH DESTINATION %arg2>
+///                }
+///             }
+///   return %res#1
+///
+///  OUTPUT:
+///   %res:3 = scf.forall ... shared_outs(%new_arg0 = %b)
+///            {
+///                ...
+///                <SOME USE OF %a>
+///                <SOME USE OF %new_arg0>
+///                <SOME USE OF %c>
+///                ...
+///                scf.forall.in_parallel {
+///                    <STORE OP WITH DESTINATION %new_arg0>
+///                }
+///             }
+///   return %res
+///
+/// NOTE: 1. All uses of the folded shared_outs (iter argument) within the
+///          scf.forall is replaced by their corresponding operands.
+///       2. The canonicalization assumes that there are no <STORE OP WITH
+///          DESTINATION *> ops within the body of the scf.forall except within
+///          scf.forall.in_parallel terminator.
+///       3. The order of the <STORE OP WITH DESTINATION *> can be arbitrary
+///          within scf.forall.in_parallel - the code below takes care of this
+///          by traversing the uses of the corresponding iter arg.
+///
+/// Example of second case :-
+///  INPUT:
+///   %res:2 = scf.forall ... shared_outs(%arg0 = %a, %arg1 = %b)
+///            {
+///                ...
+///                <SOME USE OF %arg0>
+///                <SOME USE OF %arg1>
+///                ...
+///                scf.forall.in_parallel {
+///                    <STORE OP WITH DESTINATION %arg1>
+///                }
+///             }
+///   return %res#0, %res#1
+///
+///  OUTPUT:
+///   %res = scf.forall ... shared_outs(%new_arg0 = %b)
+///            {
+///                ...
+///                <SOME USE OF %a>
+///                <SOME USE OF %new_arg0>
+///                ...
+///                scf.forall.in_parallel {
+///                    <STORE OP WITH DESTINATION %new_arg0>
+///                }
+///             }
+///   return %a, %res
+struct ForallOpIterArgsFolder : public OpRewritePattern<ForallOp> {
+  using OpRewritePattern<ForallOp>::OpRewritePattern;
+
+  /// Utility function that checks if a candidate value satisifies any of the
+  /// conditions (see above doc comment) to make it viable for folding away.
+  static bool isCandidateValueToDelete(Value result, BlockArgument blockArg) {
+    if (result.use_empty()) {
+      return true;
+    }
+    Value::user_range users = blockArg.getUsers();
+    return llvm::all_of(users, [&](Operation *user) {
+      return !isa<SubsetInsertionOpInterface>(user);
```

Comment:
We currently do not require subset insertion ops to implement the `SubsetInsertionOpInterface`. There may be ops that are inserting at a subset but do not implement the interface. Two examples:
- Unregistered ops
- `tensor.cast`. It does not implement the `SubsetInsertionOpInterface`, but the result of the cast may be passed into a subset insertion op.

So checking for `!isa<SubsetInsertionOpInterface>` is not sufficient. (You could check that *all* users implement the subset extraction op interface. Ops currently cannot be extraction and insertion ops at the same time.)

But I think there's a simpler solution: if a shared_outs bbArg is not used as the "destination" of an op inside of the `scf.forall.in_parallel` terminator, it should be safe to use the init value inside the loop instead. Can you give that a try?

Reply:
Hi @matthias-springer . Thanks for the review and suggestion! I'm writing below a few corner case which I feel might arise. Please let me know where am I going wrong with my understanding and what's the best way to deal with the same and I shall do that. :)

> There may be ops that are inserting at a subset but do not implement the interface.

Oh okay. I wasn't aware of this.
Regarding `tensor.cast` - the check I've added would only return true if at least one `SubsetInsertionOpInterface` is found for the bbArg. Even if `tensor.cast`'s result may be passed into a `SubsetInsertionOpInterface`, it won't have the same bbArg (the element type would be different), so it would be okay.

> But I think there's a simpler solution: if a shared_outs bbArg is not used as the "destination" of an op inside of the scf.forall.in_parallel terminator, it should be safe to use the init value inside the loop instead. Can you give that a try?

That is definitely correct and a simpler check, but I was trying to address the following case as well :-

```
scf.forall ... shared_outs(%arg0 = %a)
{
    ...
    <SOME USE OF %arg0>
    ...
    %x = tensor.insert_slice <some_val> into %arg0 (or some unregistered op that inserts a value into a subset)
    ...
    <some use of %x>
    ...
    scf.forall.in_parallel {
          <STORE OP WITH DESTINATION %arg0 or some other bbArg> (currently `tensor.parallel_insert_slice` and `tensor.insert_slice` do that)
    }
}
```

Therefore, if I'm only checking within `scf.forall.in_parallel`, it won't cater to the above case.

So, three things unknown to me at this point are :-
1. How to deal with the unregistered ops which might appear anywhere in the loop body and is inserting a value in the bbArg?
2. Is there any other way to check if a value is used as a "destination" of an op besides `SubsetInsertionOpInterface` ?
3. Why do we not require subset insertion ops to implement the `SubsetInsertionOpInterface` ?

Reply:
I dont know if we can get to a general solution right away. Its hard to handle the generality of all of possible "insertion-like" ops. I would go the other way. To start with, I would check that the only use of the iter_args is in `tensor.insert_in_parallel` ops within the `scf.forall.in_parallel` and only then drop the result (and corresponding iter_arg and `tensor.insert_in_parallel`. We dont have any semantics for any other case right now and not worth generalizing in a vaccum. Lets start small but well-defined and go from there

Reply:
> How to deal with the unregistered ops which might appear anywhere in the loop body and is inserting a value in the bbArg?

If you just look for the iter_arg being used as a destination in the terminator, it does not matter what the remaining loop body looks like. If there is an insertion into a tensor that is defined outside of the loop, then One-Shot Bufferize will allocate a thread-local buffer copy.

> Is there any other way to check if a value is used as a "destination" of an op besides SubsetInsertionOpInterface ?

I think there is a special interface for ops that can appear in the in_parallel terminator region. You could query that interface. Hopefully the destination can be queried from it. If not, you can add an interface method for that. For the moment you could also just hard-code the implementation to parallel_insert_slice because that's the only terminator that we support at the moment anyway.

> Why do we not require subset insertion ops to implement the SubsetInsertionOpInterface ?

One reason is that we could not handle unregistered ops correctly. Maybe there's a way to support that safely… it’s kind of like the MemoryEffectsOpInterface: if an op does not implement that interface, it does not mean that there is no side effect; it just means that we don’t know the side effects.

Assuming we only support parallel_insert_slice, does that handle all cases that you were thinking of? (I think it will work in the example that you posted.)